arch/x86/kvm/vmx/vmx.c

```diff
@@ -4734,13 +4738,60 @@ static int vmx_alloc_ipiv_pid_table(struct kvm *kvm)
 
 	kvm_vmx->pid_table = (void *)page_address(pages);
 	return 0;
+#else
+	struct kvm *shared_kvm = to_pkvm(kvm)->shared_kvm;
+	struct kvm_vmx *kvm_vmx = to_kvm_vmx(kvm);
+	struct kvm_vmx *shared_kvm_vmx;
+	u64 *pid_table;
+	int ret;
+
+	if (!irqchip_in_kernel(shared_kvm) || !enable_ipiv)
+		return 0;
+
+	pkvm_spin_lock(&to_pkvm(kvm)->lock);
+
+	if (kvm_vmx->pid_table) {
+		/* The pid_table is already set. */
+		ret = 0;
+		goto unlock;
+	}
+
+	shared_kvm_vmx = to_kvm_vmx(shared_kvm);
+
+	if (shared_kvm->arch.max_vcpu_ids > KVM_MAX_VCPU_IDS ||
+	    !shared_kvm_vmx->pid_table) {
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	kvm->arch.irqchip_mode = shared_kvm->arch.irqchip_mode;
+	kvm->arch.max_vcpu_ids = shared_kvm->arch.max_vcpu_ids;
+	pid_table = kern_pkvm_va(shared_kvm_vmx->pid_table);
+	/*
+	 * Although the contents in pid_table is not secret which is constructed
```

Comment:
nit: s/which is constructed/since it is constructed/ ?

Otherwise it sounds like a secret is constructed following the SDM but the contents in pid_table is not that secret. :)